src/V3Randomize.cpp

```diff
@@ -123,17 +142,186 @@ class RandomizeMarkVisitor final : public VNVisitorConst {
 };
 
+//######################################################################
+// Visitor that turns constraints into template strings for solvers
+
+class ConstraintExprVisitor final : public VNVisitor {
+    // NODE STATE
+    //  AstVar::user4()         -> bool. Handled in constraints
+    //  AstNodeExpr::user1()    -> bool. Depending on a randomized variable
+    // VNUser4InUse    m_inuser4;      (Allocated for use in RandomizeVisitor)
+
+    AstTask* m_taskp;
+    AstVar* m_genp;
+
+    void editFormat(AstNodeExpr* nodep) {
+        auto* const newp = new AstSFormatF{nodep->fileline(), "#b%b", false, nullptr};
+        nodep->replaceWith(newp);
+        newp->addExprsp(nodep);
+    }
+
+    // VISITORS
+    void visit(AstNodeVarRef* nodep) override {
+        if (!nodep->user1()) {
+            editFormat(nodep);
+            return;
+        }
+
+        nodep->replaceWith(new AstSFormatF{nodep->fileline(), nodep->name(), false, nullptr});
+
+        auto* const varp = nodep->varp();
+
+        VL_DO_DANGLING(pushDeletep(nodep), nodep);
+
+        if (!varp->user4()) {
+            varp->user4(true);
+            auto* const methodp = new AstCMethodHard{
+                varp->fileline(), new AstVarRef{varp->fileline(), m_genp, VAccess::READWRITE},
+                "write_var"};
+            methodp->dtypeSetVoid();
+            methodp->addPinsp(new AstVarRef{varp->fileline(), varp, VAccess::WRITE});
+            methodp->addPinsp(new AstConst{varp->dtypep()->fileline(), AstConst::Unsized64{},
+                                           (size_t)varp->width()});
+            AstNodeExpr* const varnamep
+                = new AstCExpr{varp->fileline(), "\"" + varp->name() + "\"", varp->width()};
+            varnamep->dtypep(varp->dtypep());
+            methodp->addPinsp(varnamep);
+            m_taskp->addStmtsp(new AstStmtExpr{varp->fileline(), methodp});
+        }
+    }
+    void visit(AstExtend* nodep) override {
+        if (!nodep->user1()) {
+            editFormat(nodep);
+            return;
+        }
+        int ext = nodep->width() - nodep->lhsp()->width();
+        iterateChildren(nodep);
+        AstNodeExpr* lhsp = nodep->lhsp()->unlinkFrBack();
+        AstNode* const newp = new AstSFormatF{
+            nodep->fileline(), "((_ zero_extend " + cvtToStr(ext) + ") %@)", false, lhsp};
+        nodep->replaceWith(newp);
+        VL_DO_DANGLING(pushDeletep(nodep), nodep);
+    }
+    void visit(AstNodeBiop* nodep) override {
+        if (!nodep->user1()) {
+            editFormat(nodep);
+            return;
+        }
+        iterateChildren(nodep);
+        AstNodeExpr* const lhsp = nodep->lhsp()->unlinkFrBack();
+        AstNodeExpr* const rhsp = nodep->rhsp()->unlinkFrBack();
+        std::string op;
+        switch (nodep->type()) {
+        case VNType::atLogEq:
+        case VNType::atEqWild:
+        case VNType::atEq: op = "="; break;
+        case VNType::atGt: op = "bvugt"; break;
```

Comment:
Do you think it would be worth it to go even further, making AstExtend return `"(_ zero_extend " + cvtToStr(width() - lhsp()->width()) + ")"`, or even adding some `emitSMT2` alongside `emitC` and `emitVerilog` to all AstNodeExprs?